File: apps/admin-web/src/services/userService.ts
import api from './api';
import { User, UserListResponse, UpdateUserRequest } from '../types/user';

class UserService {
  /**
   * 获取用户列表
   */
  async getUsers(page: number = 1, pageSize: number = 20, filters?: {
    status?: string;
    role?: string;
  }): Promise<UserListResponse> {
    const params: any = { page, page_size: pageSize };
    if (filters?.status) params.status = filters.status;
    if (filters?.role) params.role = filters.role;
    
    return api.get('/admin/users', { params });
  }

  /**
   * 搜索用户
   */
  async searchUsers(keyword: string, page: number = 1, pageSize: number = 20): Promise<UserListResponse> {
    return api.get('/users/search', {
      params: { keyword, page, page_size: pageSize }
    });
  }

  /**
   * 获取用户详情
   */
  async getUserById(id: number): Promise<User> {
    return api.get(`/users/${id}`);
  }

  /**
   * 添加用户
   */
  async createUser(data: {
    username: string;
    password: string;
    phone?: string;
    email?: string;
    role?: string;
  }): Promise<{ user: User }> {
    return api.post('/admin/users', data);
  }

  /**
   * 更新用户信息
   */
  async updateUser(id: number, data: UpdateUserRequest): Promise<User> {
    return api.put(`/admin/users/${id}`, data);
  }

  /**
   * 删除用户
   */
  async deleteUser(id: number): Promise<void> {
    return api.delete(`/admin/users/${id}`);
  }

  /**
   * 封禁用户
   */
  async banUser(id: number, reason: string): Promise<void> {
    return api.post(`/admin/users/${id}/ban`, { reason });
  }

  /**
   * 解封用户
   */
  async unbanUser(id: number): Promise<void> {
    return api.post(`/admin/users/${id}/unban`);
  }

  /**
   * 导出用户数据
   */
  async exportUsers(): Promise<Blob> {
    return api.get('/admin/users/export', {
      responseType: 'blob'
    });
  }
}

export default new UserService();

File: apps/backend/internal/service/user_service.go
package service

import (
	"errors"
	"github.com/lanxin/im-backend/internal/dao"
	"github.com/lanxin/im-backend/internal/model"
)

type UserService struct {
	userDAO *dao.UserDAO
	logDAO  *dao.OperationLogDAO
}

func NewUserService() *UserService {
	return &UserService{
		userDAO: dao.NewUserDAO(),
		logDAO:  dao.NewOperationLogDAO(),
	}
}

// GetUserByID 根据ID获取用户
func (s *UserService) GetUserByID(id uint) (*model.User, error) {
	return s.userDAO.GetByID(id)
}

// UpdateUser 更新用户信息
func (s *UserService) UpdateUser(userID uint, updates map[string]interface{}, ip, userAgent string) error {
	user, err := s.userDAO.GetByID(userID)
	if err != nil {
		return err
	}

	// 记录变更前的值
	oldValues := make(map[string]interface{})
	if username, ok := updates["username"]; ok {
		oldValues["username"] = user.Username
		user.Username = username.(string)
	}
	if avatar, ok := updates["avatar"]; ok {
		oldValues["avatar"] = user.Avatar
		user.Avatar = avatar.(string)
	}
	if phone, ok := updates["phone"]; ok {
		oldValues["phone"] = user.Phone
		phoneStr := phone.(string)
		if phoneStr == "" {
			user.Phone = nil
		} else {
			user.Phone = &phoneStr
		}
	}
	if email, ok := updates["email"]; ok {
		oldValues["email"] = user.Email
		emailStr := email.(string)
		if emailStr == "" {
			user.Email = nil
		} else {
			user.Email = &emailStr
		}
	}

	err = s.userDAO.Update(user)

	// 记录操作日志
	details := map[string]interface{}{
		"user_id": userID,
		"changes": map[string]interface{}{
			"old": oldValues,
			"new": updates,
		},
	}

	result := model.ResultSuccess
	errorMsg := ""
	if err != nil {
		result = model.ResultFailure
		errorMsg = err.Error()
	}

	s.logDAO.CreateLog(dao.LogRequest{
		Action:       model.ActionUserProfileUpdate,
		UserID:       &userID,
		IP:           ip,
		UserAgent:    userAgent,
		Details:      details,
		Result:       result,
		ErrorMessage: errorMsg,
	})

	return err
}

// SearchUsers 搜索用户
func (s *UserService) SearchUsers(keyword string, page, pageSize int) ([]model.User, int64, error) {
	return s.userDAO.Search(keyword, page, pageSize)
}

// ListUsers 获取用户列表
func (s *UserService) ListUsers(page, pageSize int, filters map[string]interface{}) ([]model.User, int64, error) {
	return s.userDAO.List(page, pageSize, filters)
}

// BanUser 封禁用户（管理员操作）
func (s *UserService) BanUser(adminID, targetUserID uint, reason, ip, userAgent string) error {
	user, err := s.userDAO.GetByID(targetUserID)
	if err != nil {
		return err
	}

	if user.Status == "banned" {
		return errors.New("user already banned")
	}

	oldStatus := user.Status
	user.Status = "banned"
	err = s.userDAO.Update(user)

	// 记录管理员操作日志
	details := map[string]interface{}{
		"target_user_id":   targetUserID,
		"target_username":  user.Username,
		"reason":           reason,
		"status_change":    map[string]string{"from": oldStatus, "to": "banned"},
	}

	result := model.ResultSuccess
	errorMsg := ""
	if err != nil {
		result = model.ResultFailure
		errorMsg = err.Error()
	}

	s.logDAO.CreateLog(dao.LogRequest{
		Action:       model.ActionAdminUserBan,
		AdminID:      &adminID,
		IP:           ip,
		UserAgent:    userAgent,
		Details:      details,
		Result:       result,
		ErrorMessage: errorMsg,
	})

	return err
}

